gvfs-helper-client.c

```diff
@@ -0,0 +1,378 @@
+#include "cache.h"
+#include "commit.h"
+#include "argv-array.h"
+#include "trace2.h"
+#include "progress.h"
+#include "oidset.h"
+#include "revision.h"
+#include "list-objects.h"
+#include "list-objects-filter.h"
+#include "list-objects-filter-options.h"
+#include "object.h"
+#include "object-store.h"
+#include "bisect.h"
+#include "gvfs-helper-client.h"
+#include "sub-process.h"
+#include "sigchain.h"
+#include "pkt-line.h"
+#include "quote.h"
+#include "packfile.h"
+
+static struct oidset ghc__oidset_queued = OIDSET_INIT;
+static unsigned long ghc__oidset_count;
+
+struct ghs__process {
+	struct subprocess_entry subprocess; /* must be first */
+	unsigned int supported_capabilities;
+};
+
+static int ghs__subprocess_map_initialized;
+static struct hashmap ghs__subprocess_map;
+
+#define CAP_GET      (1u<<1)
+
+static int ghc__start_fn(struct subprocess_entry *subprocess)
+{
+	static int versions[] = {1, 0};
+	static struct subprocess_capability capabilities[] = {
+		{ "get", CAP_GET },
+		{ NULL, 0 }
+	};
+
+	struct ghs__process *entry = (struct ghs__process *)subprocess;
+
+	return subprocess_handshake(subprocess, "gvfs-helper", versions,
+				    NULL, capabilities,
+				    &entry->supported_capabilities);
+}
+
+/*
+ * Send:
+ *
+ *     get LF
+ *     (<hex-oid> LF)*
+ *     <flush>
+ *
+ */
+static int ghc__get__send_command(struct child_process *process)
+{
+	struct oidset_iter iter;
+	struct object_id *oid;
+	int err;
+
+	/*
+	 * We assume that all of the packet_ routines call error()
+	 * so that we don't have to.
+	 */
+
+	err = packet_write_fmt_gently(process->in, "get\n");
+	if (err)
+		return err;
+
+	oidset_iter_init(&ghc__oidset_queued, &iter);
+	while ((oid = oidset_iter_next(&iter))) {
+		err = packet_write_fmt_gently(process->in, "%s\n",
+					      oid_to_hex(oid));
+		if (err)
+			return err;
+	}
+
+	err = packet_flush_gently(process->in);
+	if (err)
+		return err;
+
+	return 0;
+}
+
+/*
+ * Update the loose object cache to include the newly created
+ * object.
+ */
+static void ghc__update_loose_cache(const char *line)
+{
+	const char *v1_oid;
+	const char *v2_path;
+	struct object_id oid;
+	struct object_directory *odb;
+
+	if (!skip_prefix(line, "loose ", &v1_oid))
+		BUG("update_loose_cache: invalid line '%s'", line);
+
+	if (parse_oid_hex(v1_oid, &oid, &v2_path))
+		BUG("update_loose_cache: invalid oid in line '%s'", line);
+
+	while (*v2_path == ' ')
+		v2_path++;
+
+	prepare_alt_odb(the_repository);
+	for (odb = the_repository->objects->odb; odb; odb = odb->next)
```

Comment:
@jeffhostetler I think this API could be done significantly differently and make our lives less error prone. The client should pick an ODB based on the config value (and do a platform path conversion to get into some “canonical” form) and then start the helper with that path in mind. The helper could just return a loose object oid or a pack name. We wouldn’t need this loop on every response. We could navigate directly to the odb pointer we have already. 

That would at least help the windows bugs I’m having. The Mac test failure needs some more investigation.

Reply:
I like that.  Let me take a look at that.  That will simplify things considerably.